test/e2e/controller_test.go

```diff
@@ -0,0 +1,160 @@
+package e2e
+
+import (
+	"context"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/operator-framework/combo/api/v1alpha1"
+	comboConditions "github.com/operator-framework/combo/pkg/conditions"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var _ = Describe("Combination controller", func() {
+	validTemplateCR := v1alpha1.Template{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "validtemplate",
+		},
+		Spec: v1alpha1.TemplateSpec{
+			Body:       "---\nFIRSTNAME: LASTNAME",
+			Parameters: []string{"FIRSTNAME, LASTNAME"},
+		},
+	}
+
+	validCombinationCR := v1alpha1.Combination{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "validcombination",
+		},
+		Spec: v1alpha1.CombinationSpec{
+			Arguments: []v1alpha1.Argument{
+				{
+					Key: "FIRSTNAME",
+					Values: []string{
+						"John",
+						"Luke",
+					},
+				},
+				{
+					Key: "LASTNAME",
+					Values: []string{
+						"Snow",
+						"Skywalker",
+					},
+				},
+			},
+		},
+	}
+
+	expectedEvaluations := []string{
+		"John: Snow",
+		"John: Skywalker",
+		"Luke: Snow",
+		"Luke: Skywalker",
+	}
+
+	When("given healthy input and a healthy template", func() {
+		var ctx context.Context
+		var validTemplateCRCopy *v1alpha1.Template
+		var validCombinationCRCopy *v1alpha1.Combination
+
+		BeforeEach(func() {
+			ctx = context.Background()
+
+			// Create copies of valid CR's
+			validTemplateCRCopy = validTemplateCR.DeepCopy()
+			validCombinationCRCopy = validCombinationCR.DeepCopy()
+
+			// Create and defer deletion of a valid template
+			err := kubeclient.Create(ctx, validTemplateCRCopy)
+			if err != nil {
+				Fail("Failed to create validTemplateCR: " + err.Error())
+			}
```

Comment:
Making this comment generally as there's quite a few spots where this is applicable:

```
err := kubebuilder.Create(ctx, validTemplateCRCopy)
Expect(err).To(BeNil(), "failed to create validTemplateCR")
```

Reply:
Basically: use gomega assertions instead of conditional checking variables (e.g. is err nil?).

Reply:
Oh nice, way cleaner this way.